### ISSUE TITLE
Cannot define inherent `impl` on foreign type

### ISSUE THREAD

**Issue by feois** [rust-lang/rust-analyzer#18571]
**rust-analyzer version**: 0.3.2196-standalone

**rustc version**: (eg. output of `rustc -V`)

**editor or extension**: v0.3.2196

**relevant settings**: all unset

**repository link (if public, optional)**: [rust-enumeration](https://github.com/feois/rust-enumeration)

**code snippet to reproduce**: I do not know how to make a minimized reproducible example. tests/test.rs reports an error with "cannot define inherent `impl` on foreign type" and "unexpected token in input" from the macro invocation in line 56~60. However, `cargo test` runs fine without reporting any error. While attempting to reproduce the issue, you may need to modify the file (e.g. enter a new line and delete it) as the error may not be reported. Attempting to inline macro will result in missing tokens. I do not know if this is only reproducible on my machine.

**Comment by feois** [comment]
The error reported now changes to a compile error only raised by my macro in the latest commit, however I am pretty sure (as the author of the macro) that the error should not be raised, evidenced by the fact that `cargo test` or `cargo check` both do not report any errors.

[Old commit](https://github.com/feois/rust-enumeration/tree/cf5a0de348b779e9d9a62351335a7165becfa6b2)
[Latest commit (by the time of writing)](https://github.com/feois/rust-enumeration/tree/773992b502e323061684b27daefa61228fda2986)

I am now pretty certain this is an error in rust-analyzer's macro expansion, thus I am changing the issue title.

**Comment by flodiebold** [comment]
You can use the "Rust Analyzer: Expand macro recursively" function to see what RA expands the macro to and try to find what's wrong there.

**Comment by flodiebold** [comment]
You know your macro though, so hopefully could be able to tell what part of the expansion is wrong? 😅

**Comment by brodiealexander** [comment]
I've encountered this issue myself. I have a macro that converts the input TokenStream to a string and then does naive string replacement for `#ty` etc. 

The rust-analyzer recursive expansion seems to inject spaces or newlines between the # and the span after it:

(The input)
```Rust
enum_generic_for_each_variant!(GenericStriped,StripedBuffer,(u8, u16, u32, u64, i8, i16, i32, i64, f32, f64),
    {
        impl From<#inner<#ty>> for #outer {
            fn from(value: #inner<#ty>) -> Self {
                #outer::#ty_pc(value)
            }
        }
    }
);
```

```
// Recursive expansion of enum_generic_for_each_variant! macro
// ============================================================

impl From<
#
inner<#
ty>>for #
outer {
    fn from(value: #inner<#ty>)
    ->Self {
        #
        outer::#ty_pc(value)
    }
}
```

Resulting in the foreign impl error. The code does, however, compile without issue. 

Doing some filtering to remove all whitespace except ASCII space and then replacing "# " with "#" seems to make rust-analyzer happy
